Eastern-Research-Group/mywaterway
# Bugfix/107 legend issues

## Related Issues:
* https://jira.epa.gov/browse/HMW-107
* https://jira.epa.gov/browse/HMW-117 (duplicate to be deleted)

## Main Changes:
* Fixed various issues with the legend
  1. Esri "No legend" text being displayed when there are custom HMW layers being shown in the legend.
  2. HMW "There are currently no items to display" text being displayed when there are layers being shown from the Add Data Widget.
  3. HMW layers being shown in the legend, even though the map zoom level is outside of the layer's zoom threshold. 
  4. Sometimes layers are duplicated in the legend. Basically, we were displaying our custom legend for the layer and esri was displaying the out of the box version.

Implementation Note: I couldn't find a good ESRI event to hook into to see if the Esri portion of the widget had anything to display. I ended up using MutationObserver to detect when Esri adds/removes the "No legend" text to/from the DOM and used this information to know whether the Esri portion of the legend was actually displaying anything. 

## Steps To Test:
* Issue 1 and 4
  1. Navigate to http://localhost:3000/community
  2. Open the Add Data Widget
  3. Add the "USA Flood Hazard Areas" layer
  4. Open the legend
  5. Verify the legend only says "There are currently no items to display"
      * NOTE: In production this currently says "There are currently no items to display" and "No legend" (simulates issue 1)
  6. Open the layer list 
  7. Turn on the "County" layer
  8. Verify the legend only displays "County" once
      * NOTE: In production this will display "County" twice (simulates issue 4)
  9. Zoom in until the "USA Flood Hazard Areas" layer is displayed
  10. Remove the "USA Flood Hazard Areas" layer
  11. Re-add the "USA Flood Hazard Areas" layer
  12. Zoom out until the "USA Flood Hazard Areas" layer is no longer visible
  13. Open the legend
  14. Verify the legend only says "County" once
      * NOTE: In production this will display "County" and "No legend" (simulates issue 1)

## Steps To Test:
* Issue 2
  1. Navigate to http://localhost:3000/community or refresh the page
  2. Open the Add Data Widget
  3. Add the "USA Flood Hazard Areas" layer
  4. Zoom in until the "USA Flood Hazard Areas" layer is displayed
  5. Open the legend
  6. Verify the legend only displays info for the "USA Flood Hazard Areas" layer
      * NOTE: In production this will display "There are currently no items to display" and "USA Flood Hazard Areas" layer info.
  
## Steps To Test:
* Issue 3
  1. Navigate to http://localhost:3000/community/dc/overview
  2. Turn off the "Waterbodies" layer and leave the "Surrounding Waterbodies" layer on
  3. Zoom out until the "Surrounding Waterbodies" layer is no longer visible
  4. Open the legend 
  5. Verify the legend only displays "Searched Location" and "HUC12 Boundaries"
